File: 303_sum_range/sum_range.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-03-01 20:29:42
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-03-01 20:40:26
 */

struct NumArray {
    pre_sum: Vec<i32>,
}

impl NumArray {
    fn new(nums: Vec<i32>) -> Self {
        let mut pre = 0;
        let pre_sum: Vec<i32> = nums
            .iter()
            .map(|x| {
                pre += x;
                pre
            })
            .collect();

        NumArray { pre_sum: pre_sum }
    }

    fn sum_range(&self, i: i32, j: i32) -> i32 {
        if i > j {
            return 0;
        }
        if i == 0 {
            return self.pre_sum[j as usize];
        }

        return self.pre_sum[j as usize] - self.pre_sum[i as usize - 1];
    }
}

fn main() {
    let obj = NumArray::new(vec![-2, 0, 3, -5, 2, -1]);
    assert_eq!(obj.sum_range(0, 2), 1);
    assert_eq!(obj.sum_range(2, 5), -1);
    assert_eq!(obj.sum_range(0, 5), -3);
}

File: cpp/303_sum_range/sum_range.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-03-01 20:29:40
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-03-01 20:36:50
 */

#include <cassert>
#include <vector>

using namespace std;

class num_array {
 public:
  num_array(vector<int>&& nums) {
    int n = nums.size();
    sums.resize(n + 1);
    for (int i = 0; i < n; i++) {
      sums[i + 1] = sums[i] + nums[i];
    }
  }
  int sum_range(int i, int j) { return sums[j + 1] - sums[i]; }

 private:
  vector<int> sums;
};

int main() {
  num_array array{vector<int>{-2, 0, 3, -5, 2, -1}};
  assert(array.sum_range(0, 2) == 1);
  assert(array.sum_range(2, 5) == -1);
  assert(array.sum_range(0, 5) == -3);
  return 0;
}